examples/mem-correctness-test/src/main.rs

```diff
@@ -0,0 +1,279 @@
+    /* This function assumes that the percentage of free heap space is more
+     * than MIN_ALLOWED_FREE_HEAP_PERCENTAGE
+     */
+
+    let random_size = get_random_num(mem::size_of::<usize>() * 2, PER_THREAD_PER_BUFFER_MAX_SIZE)
+        / mem::size_of::<usize>();
+    let mut tmp: Vec<usize> = Vec::with_capacity(random_size);
+    tmp.resize(random_size, magic_num_to_check);
+    array_of_vectors.push(tmp);
+    update_occupied_heap_size_on_addition(random_size * mem::size_of::<usize>());
+}
+
+fn worker_thread(tid: i32, pair_clone: Arc<(Mutex<bool>, Condvar)>, magic_num_to_check: usize) {
+    /* Wait for all the threads to be created and then start together */
+    wait_per_thread(pair_clone);
+
+    let mut array_of_vectors: Vec<Vec<usize>> = Vec::new();
+
+    /* Once the thread's allocation and deallocation operations begin, we
+     * shouldn't take any lock as the allocator that we trying to test is a
+     * multithreaded allocator and we should allow as many threads as possible
+     * to get the lock.
+     */
+    for _i in 1..=MAX_ITERATIONS_PER_THREAD {
+        let ran_choice = get_random_num(0, NUM_OPERATION_CHOICES - 1);
+
+        match ran_choice {
+            0 => {
+                select_and_check_random_buffer_contents(&array_of_vectors, magic_num_to_check);
+                println!("T-{} check", tid);
+            }
+            1..=2 => {
+                /* Although  get_free_heap_percentage() is thread safe, this
+                 * match case may not be completely thread safe as we are only
+                 * interested in an approximate value of the remaining free space
+                 * percentage.
+                 */
+                if get_free_heap_percentage() > MIN_ALLOWED_FREE_HEAP_PERCENTAGE {
+                    add_new_buffer(&mut array_of_vectors, magic_num_to_check);
+                    println!("T-{} allocate", tid);
+                } else {
+                    println!("T-{} SKIP", tid);
+                }
+            }
+            3 => {
+                delete_random_buffer(&mut array_of_vectors);
+                println!("T-{} delete", tid);
+            }
+            _ => {
+                panic!("Invalid random operation choice done");
+            }
+        }
+    }
+}
+
+fn spawn_threads(thread_count: i32, magic_num_to_check: usize) {
+    let mut handles = vec![];
+
+    let pair = Arc::new((Mutex::new(false), Condvar::new()));
+    for i in 0..thread_count {
+        // Spawn a thread that waits until the condition is met
+        let pair_clone = Arc::clone(&pair);
+        let handle = thread::spawn(move || {
+            worker_thread(i, pair_clone, magic_num_to_check);
+        });
+        handles.push(handle);
+    }
+
+    /* Start all the threads */
+    wakeup_all_child_threads(pair);
+
+    /* Wait for all threads to finish */
+    for handle in handles {
+        handle.join().unwrap();
+    }
+}
+
+fn get_num_processors() -> usize {
```

Comment:
This is confusing as `get_num_processors` doesn't return the number of processors on the machine

Reply:
As an optimization I will fill up this function with so that it returns number of processors. Currently it returns a hardcoded value

Reply:
Yes but if it's hardcoded, there's no need to have this function